examples/simple/simple.cpp

```diff
@@ -84,41 +87,93 @@ int main(int argc, char ** argv) {
     LOG("\n");
 
     for (auto id : tokens_list) {
-        LOG("%s", llama_token_to_piece(ctx, id).c_str());
+        if (id == TOKEN_IMG_PLACEMENT) {
+            LOG("<img_placement>");
+        } else {
+            LOG("%s", llama_token_to_piece(ctx, id).c_str());
+        }
     }
 
+    LOG("\n\n");
+
+    // load image
+    llama_batch_img img_batch = llama_batch_img_init(1);
+    img_batch.imgs[0] = load_image_from_file("../models/eiffel-tower-3349075_1280.jpg");
+
     // create a llama_batch with size 512
     // we use this object to submit token data for decoding
 
     llama_batch batch = llama_batch_init(512, 0, 1);
 
     // evaluate the initial prompt
-    for (size_t i = 0; i < tokens_list.size(); i++) {
-        llama_batch_add(batch, tokens_list[i], i, { 0 }, false);
+    int n_cur = 0;
+    int i_img = 0;
+    for (auto id : tokens_list) {
+        if (id == TOKEN_IMG_PLACEMENT) {
+            img_batch.pos[i_img] = n_cur;
+            n_cur += llama_img_n_tokens(ctx, img_batch.imgs[i_img]);
+            i_img++;
+        } else {
+            llama_batch_add(batch, id, n_cur, { 0 }, false);
+            printf("pos %d tok %d --> %s\n", n_cur, id, llama_token_to_piece(ctx, id).c_str());
+            n_cur++;
+        }
     }
 
     // llama_decode will output logits only for the last token of the prompt
     batch.logits[batch.n_tokens - 1] = true;
 
+    if (llama_encode_vision(ctx, img_batch) != 0) {
+        LOG("%s: llama_encode_vision() failed\n", __func__);
+        return 1;
+    }
+
+    n_cur = 0;
+    {
+        auto t1 = ::llama_tokenize(ctx, "A chat between a curious human and an artificial intelligence assistant. The assistant gives helpful, detailed, and polite answers to the human's questions.\nUSER:", false);
+        auto t2 = ::llama_tokenize(ctx, "\nwhat did you see?\nASSISTANT:", false);
+        t1.insert(t1.begin(), 1);
+
+        n_cur = 0;
+        llama_batch_clear(batch);
+        llama_batch_add(batch, 1, 0, { 0 }, false);
+        llama_decode(ctx, batch);
+
+        n_cur = t1.size();
+        llama_batch_clear(batch);
+        llama_batch batch0 = {int32_t(576), nullptr, _test_get_img_embd(ctx), nullptr, nullptr, nullptr, nullptr, n_cur, 1, 0, };
+        llama_decode(ctx, batch0);
+
+        n_cur = 0;
+        llama_batch_clear(batch);
+        for (auto t : t1) { llama_batch_add(batch, t, n_cur, { 0 }, false); n_cur++; }
+        llama_decode(ctx, batch);
+
+        n_cur = t1.size() + 576;
+        llama_batch_clear(batch);
+        printf("pos %d\n", n_cur);
+        for (auto t : t2) { llama_batch_add(batch, t, n_cur, { 0 }, false); n_cur++; }
+        batch.logits[batch.n_tokens - 1] = true;
```

Comment:
@ggerganov I observed something weird here that I can't understand, so I would like to ask if you know why.

What I'm trying to do here is to test if I can call `llama_decode` on different parts of the sentence out-of-order. That means instead of evaluating `text,image,text`, I try evaluating `image,text,text`. In theory, this should work because the position is set correctly for all input tokens/embeddings, just the placement in KV cache is different.

But in reality, `llama_decode` on a batch containing `embd` doesn't work correctly if the **KV cache is empty**. In this example, I need to evaluate at least one token for it to work.

I'd be happy to provide more info to help pin-point the problem. Thank you.

Reply:
> But in reality, llama_decode on a batch containing embd doesn't work correctly if the KV cache is empty.

What error do you get?

Reply:
There is no error, but it the generated text is quite irrelevant.
- Expected output: `The image features a beautiful view of the Eiffel Tower`
- Without the first single-token `llama_decode`: `The image features a cityscape with a tall building prominently visible in the center. The building appears to be lit up, possibly at night or during` --> clearly wrong because I provided a daytime image

Please note that if the code block follow order of `text,image,text`, then it will work correctly (generate expected output).

I have a feeling that the KV for embedding is truncated somehow, but can't find a way to very this. I tried playing with `llama_kv_cache_view` but it seems like it's all OK.

<details>
  <summary>show code</summary>

```cpp
        llama_kv_cache_view v = llama_kv_cache_view_init(ctx, 1);
        llama_kv_cache_view_update(ctx, &v);
        printf("n_cells %d\n", v.n_cells);
        for (int i = 0; i < 16; i++) {
            printf("cell %d --> pos %d (seq %d)\n", i, v.cells[i].pos, v.cells_sequences[0]);
        }
        for (int i = 570; i < 570+16; i++) {
            printf("cell %d --> pos %d (seq %d)\n", i, v.cells[i].pos, v.cells_sequences[0]);
        }
        for (int i = n_cur-8; i < n_cur+8; i++) {
            printf("cell %d --> pos %d (seq %d)\n", i, v.cells[i].pos, v.cells_sequences[0]);
        }
```

</details>


(I'm uploading the model so that you can test it. Will drop a link here when it's available)

Reply:
Link to gguf: https://huggingface.co/ngxson/test-llava-will-be-deleted-soon/blob/main/model.gguf

Command: `./llama-simple -m ../models/llava-1.5-7b-hf/model.gguf -c 1024 -v`

Reply:
I don't think that `text,image,text` is equivalent to `image,text,text` regardless if the positions are correct. But before we discuss that, it seems to me that the positions are off by one due to the BOS token:

```c++
        n_cur = 0;
        llama_batch_clear(batch);
        llama_batch_add(batch, 1, 0, { 0 }, false);
        llama_decode(ctx, batch);

        n_cur = t1.size();
        llama_batch_clear(batch);
        llama_batch batch0 = {int32_t(576), nullptr, _test_get_img_embd(ctx), nullptr, nullptr, nullptr, nullptr, 1 + n_cur, 1, 0, };
        llama_decode(ctx, batch0);

        n_cur = 0;
        llama_batch_clear(batch);
        for (auto t : t1) { llama_batch_add(batch, t, 1+ n_cur, { 0 }, false); n_cur++; }
        llama_decode(ctx, batch);

        n_cur = t1.size() + 576;
        llama_batch_clear(batch);
        printf("pos %d\n", n_cur);
        for (auto t : t2) { llama_batch_add(batch, t, 1+ n_cur, { 0 }, false); n_cur++; }
        batch.logits[batch.n_tokens - 1] = true;
```

Reply:
The BOS is added manually, `t1.insert(t1.begin(), 1);`

The `text,image,text` sequence looks like this:

```cpp
        auto t1 = ::llama_tokenize(ctx, "A chat between a curious human and an artificial intelligence assistant. The assistant gives helpful, detailed, and polite answers to the human's questions.\nUSER:", false);
        auto t2 = ::llama_tokenize(ctx, "\nwhat did you see?\nASSISTANT:", false);
        t1.insert(t1.begin(), 1);

        n_cur = 0;
        llama_batch_clear(batch);
        for (auto t : t1) { llama_batch_add(batch, t, n_cur, { 0 }, false); n_cur++; }
        llama_decode(ctx, batch);

        n_cur = t1.size();
        llama_batch_clear(batch);
        llama_batch batch0 = {int32_t(576), nullptr, _test_get_img_embd(ctx), nullptr, nullptr, nullptr, nullptr, n_cur, 1, 0, };
        llama_decode(ctx, batch0);

        n_cur = t1.size() + 576;
        llama_batch_clear(batch);
        printf("pos %d\n", n_cur);
        for (auto t : t2) { llama_batch_add(batch, t, n_cur, { 0 }, false); n_cur++; }
        batch.logits[batch.n_tokens - 1] = true;
```

By simply moving the code block `llama_decode(ctx, batch0);` to the beginning, you can observe the different (even though all token positions stays the same)

Reply:
> But before we discuss that, it seems to me that the positions are off by one due to the BOS token

Just to clarify a bit, the single-token llama_decode is there just for testing. Indeed, adding that will effective make the sequence to have 2 BOS at the same position 0. But for some reasons, the output is the same with single BOS: `bos,text,image,text`

Reply:
Thanks for your response. FYI, I tried printing the `inp_KQ_mask` and turns out it's the problem. In the second case `image, text0, text1`, somehow the `text1` can't attend to tokens in `text1`.

What's weird is that, if now I have 4 batches: `bos,image,text0,text1`, then it will work correctly (`bos` is the batch that contains one single BOS token)

I also set the `n_batch` and `n_ubatch` to 1024 to make sure the image fit into one batch, but it doesn't solve my issue.

> One thing to be careful is when computing the `image` batch, what mask should we use?

In the case of llava model, let's imagine that the image batch is just a batch containing 576 tokens, so let's call it `textI` instead of `image`. So my assumption is `text0,textI,text1` should be equivalent to `textI,text0,text1` (?). When generating the next token, we simply add all tokens that we already evaluated to the mask, so it should be a auto-aggressive mask.

Perhaps PaliGemma is different because it requires non-causal.

---

I also tried setting `llama_set_causal_attn` to false before evaluating `image` batch, but that doesn't help either.

Reply:
For now, I think a hacky way would be to internally populate image tokens into the batch in `llama_decode`. I'm also thinking about possibility of extending `build_inp_pos` to accept both token and embedding input at the same time. What do you think? (also cc @abetlen since that what your #7553 is about)